gitpod-io/leeway
# Only download remote packages when required for build

## Description


This changes Leeway so that it only downloads packages form the remote cache when they're needed to perform the actual build.

For gitpo-io/gitpod this means that `leeway run dev/preview:build` in a fresh workspace goes from ~4m to 40s.

**Still to be considered**

1. I still need to implement the Exists function for minio
1. This currently no longer supports the "additional caches" feature of `--add-remote-cache`. Is it worth extending this PR to support that or should we remove the feature? We don't currently use that feature in our gitpod-io/gitpod build 

## Related Issue(s)

Fixes https://github.com/gitpod-io/ops/issues/6255

## How to test


Besides the unit tests, I tested this on our build in gitpod-io/gitpod.

Open a workspace in gitpod-io/gitpod and download the version of Leeway from this branch:

```sh
go install github.com/gitpod-io/leeway@05bb55418a58f7bb9f0f27bd3fad0fa79d1333b5
which leeway # should be /workspace/go/bin/leeway
```

Now perform a build (you'll have to remote the `--dont-retag` from `build.sh` manually for now.

```sh
rm -rf /tmp/cache && rm -rf /tmp/build # ensure you have a clean cache etc.
leeway run dev/preview:build
```

Perform a subsequent build to verify it checks the remote caches for fewer packages, and that it doesn't try to download any new packages.

```sh
leeway run dev/preview:build
```

## Release Notes

TBD


```release-note
NONE
```

## Documentation


TBD
